README.md
<h1 align="center">Sniper</h1>

<p align="center">
  <img src="https://www.codefactor.io/repository/github/slow/nitro-sniper/badge" /> 
  <img src="https://img.shields.io/github/issues/slow/nitro-sniper?style=flat" />
  <img src="https://img.shields.io/github/stars/slow/nitro-sniper?style=flat" />
</p>

<p align="center">
   A lightweight, fast and efficient discord nitro sniper, giveaway sniper & invite sniper. 
   <img src="https://media.wtf/11603354" />
</p>

# Features
- Light, fast & efficient
- Nitro sniper
- Giveaway sniper
- Limit nitro redeems (e.g. 2 codes / 24 hours)
- Custom giveaway reaction delay (randomized between **delay** and (**delay** * 1.5))
- DM host on giveaway win
- Custom DM Delay
- Custom DM Message
- One-click Deploy
- Duplicate Protection
- Fake Code Detection
- Multi-Account Support
- 3 Modes - `main`, `alts`, `both`
- Webhook support (Giveaways & Nitro)
- Fully customizable webhook
- Ability to turn the webhook on/off for certain events
- @everyone ping on webhook message
- Ability to turn the ping on/off for certain events
- Whitelist words for the giveaway sniper
- Blacklist words for the giveaway sniper
- Whitelisted words only mode for giveaway sniper (Blacklist still applies)
- Blacklist servers from the giveaway sniper
- Announce nitro type on snipe
- Invite Sniper
- Minimum member count for invite sniper
- Maximum member count for invite sniper
- Limit invite joins (e.g. 10 invites / 6 hours)
- Minimum delay before sniping invite
- Maximum delay before sniping invite

## Installation
https://github.com/slow/nitro-sniper/wiki/Installation

## Default Configuration
```js
{
   "tokens": {
      "main": "Nz...", // Main Token
      "alts": [
         "Nz..." // Alt Tokens
      ],
   },
   "mode": "both", // The mode to run the sniper in. Options: main (only main account), alts (only alts), both
   "nitro": {
      "max": 2, // The amount of nitros needed to be sniped for the cooldown to activate
      "cooldown": 24 // Cooldown to activate after max nitro has been hit (in hours)
   },
   "giveaway": {
      "enabled": true, // Wether or not to activate the giveaway sniper (true/false)
      "delay": 30, // Delay to react to the giveaway (in seconds)
      "dm": true, // DM the hoster on giveaway win (true/false)
      "dmMessage": "Hey, i won the giveaway. Could i redeem my prize?", // Message to DM the host
      "dmDelay": 25, // How long to wait to DM (in seconds)
      "blacklistedWords": [ // Blacklisted words for giveaway prizes
         "bot",
         "test",
         "ban"
      ],
      "whitelistOnly": false, // Only react to whitelisted giveaway prizes (true/false)
      "whitelistedWords": [ // Whitelisted words for giveaway prizes
         "nitro"
      ],
      "blacklistedServers": [ // Blacklisted Server IDs to not snipe giveaways on
         ""
      ]
   },
   "invite": {
      "enabled": false, // Wether or not to activate the invite sniper (true/false)
      "delay": {
         "min": 10, // Minimum delay to join the server (in seconds)
         "max": 20 // Maximum delay to join the server (in seconds)
      },
      "members": {
         "min": 1500, // The minimum member count the server should have
         "max": 50000 // The maximum member count the server should have
      },
      "max": 10, // The amount of joined invites needed for the cooldown to activate
      "cooldown": 6 // Cooldown to activate after max joined invites has been hit (in hours)
   },
   "webhook": { 
      "url": "https://discord.com/api/webhooks/.../...", // URL to fire webhook to for notifications
      "enabled": {
         "codeInvalid": false, // Fire webhook on invalid code (true/false)
         "codeAlreadyRedeemed": false, // Fire webhook on already redeemed code (true/false)
         "codeSuccess": true, // Fire webhook on sniped code (true/false)
         "giveawayEntered": true, // Fire webhook on giveaway enter (true/false)
         "giveawayWin": true, // Fire webhook on giveaway win (true/false)
         "inviteJoin": false // Fire webhook on invite join (true/false)
      }, 
      "mentionEveryone": { 
         "codeInvalid": false, // Mention on invalid code (true/false)
         "codeAlreadyRedeemed": false, // Mention on already redeemed code (true/false)
         "codeSuccess": true, // Mention on sniped code (true/false)
         "giveawayEntered": false, // Mention on giveaway enter (true/false)
         "giveawayWin": true, // Mention on giveaway win (true/false)
         "inviteJoin": false // Mention on invite join (true/false)
      }
   }
}
```

#### How to obtain your token
https://github.com/Tyrrrz/DiscordChatExporter/wiki/Obtaining-Token-and-Channel-IDs#how-to-get-a-user-token

# Tips
- Try to get a low latency to discord servers as there can be competition with other snipers.
- This is technically a self-bot: mentioning this in a discord chat is enough to make your account reportable to Trust & Safety.
- Running more than one instance or different snipers is an easy way to get your account deactivated.
- Before using, consider the moral implications of stealing gifts from communities you have nothing to do with.

> If you like my project, please consider starring the repo :)


CHANGELOG.md
# [1.8.0](https://github.com/slow/nitro-sniper/compare/v1.7.2...v1.8.0) (2021-02-20)


### Features

* fix the fucking heroku issue once and for all ([87ce783](https://github.com/slow/nitro-sniper/commit/87ce783901d72ff5566ef901ecc0d8fcd0f2f333))

## [1.7.2](https://github.com/slow/nitro-sniper/compare/v1.7.1...v1.7.2) (2021-02-20)


### Bug Fixes

* heroku suspension | closes [#29](https://github.com/slow/nitro-sniper/issues/29) ([cd5431f](https://github.com/slow/nitro-sniper/commit/cd5431f7001aa015b1c1121b6377d7daafe09eb4))

## [1.7.1](https://github.com/slow/nitro-sniper/compare/v1.7.0...v1.7.1) (2021-02-20)


### Bug Fixes

* dumb fuck ([6c01c14](https://github.com/slow/nitro-sniper/commit/6c01c146a6a07d94c96f6b69b9dd989796ada15e))

# [1.7.0](https://github.com/slow/nitro-sniper/compare/v1.6.6...v1.7.0) (2021-02-15)


### Features

* revert changes ([3735db2](https://github.com/slow/nitro-sniper/commit/3735db2d03ec20eef7e37c6b8bb323b4bd9c9d49))

## [1.6.2](https://github.com/slow/nitro-sniper/compare/v1.6.1...v1.6.2) (2021-02-09)


### Bug Fixes

* closes [#18](https://github.com/slow/nitro-sniper/issues/18) ([a53c6fe](https://github.com/slow/nitro-sniper/commit/a53c6fe8192b8f24be08b5baba344f83736224c3))

## [1.6.1](https://github.com/slow/nitro-sniper/compare/v1.6.0...v1.6.1) (2021-02-07)


### Bug Fixes

* low iq ([d5fb5f5](https://github.com/slow/nitro-sniper/commit/d5fb5f544cb5aea0e8966398b5785f7d618064e6))

# [1.6.0](https://github.com/slow/nitro-sniper/compare/v1.5.3...v1.6.0) (2021-02-07)


### Features

* webhook support for invite join ([bf5b38c](https://github.com/slow/nitro-sniper/commit/bf5b38c04a8bc172e2feb35c9896a57229ed7acc))

## [1.5.3](https://github.com/slow/nitro-sniper/compare/v1.5.2...v1.5.3) (2021-02-07)


### Bug Fixes

* readme type ([d2f6c79](https://github.com/slow/nitro-sniper/commit/d2f6c793b30ce9e9328289d6eedc26bcdedf6c06))

## [1.5.2](https://github.com/slow/nitro-sniper/compare/v1.5.1...v1.5.2) (2021-02-07)


### Bug Fixes

* unverified accs ([2e70cc6](https://github.com/slow/nitro-sniper/commit/2e70cc67552e733220afb854c7ae78ae3c3c16da))

## [1.5.1](https://github.com/slow/nitro-sniper/compare/v1.5.0...v1.5.1) (2021-02-07)


### Bug Fixes

* check for max servers ([0e6c21c](https://github.com/slow/nitro-sniper/commit/0e6c21c73772b9d400ce5786218efc6cdfb85df2))

# [1.5.0](https://github.com/slow/nitro-sniper/compare/v1.4.1...v1.5.0) (2021-02-07)


### Features

* invite sniper ([3955351](https://github.com/slow/nitro-sniper/commit/39553517b51ef9dcba5a6d7c7c7ccfefccd0d14b))

## [1.4.1](https://github.com/slow/nitro-sniper/compare/v1.4.0...v1.4.1) (2021-02-07)


### Bug Fixes

* handle cooldown on messages with multiple codes ([e71b238](https://github.com/slow/nitro-sniper/commit/e71b23886c005590bae8fa6051ea36b2d92b8aa2))

# [1.4.0](https://github.com/slow/nitro-sniper/compare/v1.3.1...v1.4.0) (2021-02-07)


### Features

* cacheless discord v12 upgrade ([09b283a](https://github.com/slow/nitro-sniper/commit/09b283afd12b8a5610fc67da8425678198068090))

## [1.3.1](https://github.com/slow/nitro-sniper/compare/v1.3.0...v1.3.1) (2021-02-06)


### Bug Fixes

* settings is always defined ([55795ec](https://github.com/slow/nitro-sniper/commit/55795eca862bacad866e82fd7fcb1d8107f63764))

# [1.3.0](https://github.com/slow/nitro-sniper/compare/v1.2.3...v1.3.0) (2021-02-05)


### Features

* handler for unknown responses ([c7dbe0b](https://github.com/slow/nitro-sniper/commit/c7dbe0b4154bb2d5e5c7994372963a05553db94f))

## [1.2.3](https://github.com/slow/nitro-sniper/compare/v1.2.2...v1.2.3) (2021-02-05)


### Bug Fixes

* stupid mistake, sniper will be same as before now. ([74ef3d6](https://github.com/slow/nitro-sniper/commit/74ef3d6d277660e39639ad05f48f94123220e551))

## [1.2.2](https://github.com/slow/nitro-sniper/compare/v1.2.1...v1.2.2) (2021-01-31)


### Bug Fixes

* fire webhook by default when giveaway entered ([a1a35da](https://github.com/slow/nitro-sniper/commit/a1a35da38e18d9ebfbca7b728a77c563bc68e4e1))

## [1.2.1](https://github.com/slow/nitro-sniper/compare/v1.2.0...v1.2.1) (2021-01-31)


### Bug Fixes

* update outdated giveaway react delay ([8c4892b](https://github.com/slow/nitro-sniper/commit/8c4892b393b85b6cb69eb968359902bd27537959))

# [1.2.0](https://github.com/slow/nitro-sniper/compare/v1.1.1...v1.2.0) (2021-01-31)


### Features

* customizable webhook events ([3f99823](https://github.com/slow/nitro-sniper/commit/3f99823a2fad2bb2680d448e9d30e500c2168feb))

## [1.1.1](https://github.com/slow/nitro-sniper/compare/v1.1.0...v1.1.1) (2021-01-31)


### Bug Fixes

* toLowerCase of undefined ([5ce052c](https://github.com/slow/nitro-sniper/commit/5ce052cf996a728f71bce7ce5a159f3512a596ff))

# [1.1.0](https://github.com/slow/nitro-sniper/compare/v1.0.1...v1.1.0) (2021-01-31)


### Features

* per-type customizable pings ([f09793b](https://github.com/slow/nitro-sniper/commit/f09793bdd64403dc3130d108b004ab070993c0dd))

## [1.0.1](https://github.com/slow/nitro-sniper/compare/v1.0.0...v1.0.1) (2021-01-31)


### Bug Fixes

* promise rejection errors ([3f781a5](https://github.com/slow/nitro-sniper/commit/3f781a545ac6279109a99400746dc43a928c6414))

# 1.0.0 (2021-01-31)


### Bug Fixes

* version ([3c49381](https://github.com/slow/nitro-sniper/commit/3c493816dc66bbd77e05499e1eec07df17e63a4a))

## [1.1.11](https://github.com/slow/nitro-sniper/compare/v1.1.10...v1.1.11) (2021-01-31)


### Bug Fixes

* low iq ([1dfa9f8](https://github.com/slow/nitro-sniper/commit/1dfa9f8d00317b635a1ea8e319ed7fab4fdd7f26))

## [1.1.10](https://github.com/slow/nitro-sniper/compare/v1.1.9...v1.1.10) (2021-01-31)


### Bug Fixes

* false positive - closes [#15](https://github.com/slow/nitro-sniper/issues/15) ([e5eb1c5](https://github.com/slow/nitro-sniper/commit/e5eb1c5571fa2abb73eda644e90f628694968ac7))


src/lib/Constants.js
const chalk = require('chalk');

module.exports = {
   clientOptions: {
      disabledEvents: [
         'TYPING_START',
         'VOICE_SERVER_UPDATE',
         'RELATIONSHIP_ADD',
         'RELATIONSHIP_REMOVE',
         'GUILD_ROLE_DELETE',
         'GUILD_ROLE_UPDATE',
         'GUILD_BAN_ADD',
         'GUILD_BAN_REMOVE',
         'CHANNEL_UPDATE',
         'CHANNEL_PINS_UPDATE',
         'MESSAGE_DELETE',
         'MESSAGE_UPDATE',
         'MESSAGE_DELETE_BULK',
         'MESSAGE_REACTION_ADD',
         'MESSAGE_REACTION_REMOVE',
         'GUILD_MEMBER_UPDATE',
         'GUILD_MEMBERS_CHUNK',
         'GUILD_ROLE_CREATE',
         'MESSAGE_REACTION_REMOVE_ALL',
         'USER_UPDATE', 'USER_NOTE_UPDATE',
         'USER_SETTINGS_UPDATE',
         'PRESENCE_UPDATE',
         'VOICE_STATE_UPDATE',
         'GUILD_UPDATE',
         'GUILD_MEMBER_ADD',
         'GUILD_MEMBER_REMOVE'
      ],
      messageEditHistoryMaxSize: 1,
      messageSweepInterval: 1,
      messageCacheLifetime: 1,
      messageCacheMaxSize: 1,
      cacheChannels: true,
      restTimeOffset: 0,
      presence: {
         afk: true
      },
      ws: {
         properties: {
            '$os': 'Discord Android',
            '$browser': 'Discord Android',
            '$device': 'Discord Android'
         }
      }
   },
   defaultSettings: {
      tokens: {
         main: '',
         alts: [
            ''
         ],
      },
      mode: 'both',
      nitro: {
         max: 2,
         cooldown: 24,
      },
      giveaway: {
         enabled: true,
         delay: 30,
         dm: true,
         dmMessage: 'Hey, i won the giveaway. Could i redeem my prize?',
         dmDelay: 25,
         blacklistedWords: [
            'bot',
            'test',
            'ban'
         ],
         whitelistOnly: false,
         whitelistedWords: [
            'nitro'
         ],
         blacklistedServers: [
            ''
         ]
      },
      invite: {
         enabled: false,
         delay: {
            min: 10,
            max: 20
         },
         members: {
            min: 1500,
            max: 50000
         },
         max: 10,
         cooldown: 6
      },
      webhook: {
         url: '',
         enabled: {
            codeInvalid: false,
            codeAlreadyRedeemed: false,
            codeSuccess: true,
            giveawayEntered: true,
            giveawayWin: true,
            inviteJoin: false
         },
         mentionEveryone: {
            codeInvalid: false,
            codeAlreadyRedeemed: false,
            codeSuccess: true,
            giveawayEntered: false,
            giveawayWin: true,
            inviteJoin: false
         }
      }
   },
   cooldown: (type, max, hours) => `Cooldown hit after ${max} ${type}s were sniped. Pausing the ${type} sniper for ${hours} hours.`,
   xSuperProperties: 'eyJvcyI6IldpbmRvd3MiLCJicm93c2VyIjoiRGlzY29yZCBDbGllbnQiLCJyZWxlYXNlX2NoYW5uZWwiOiJjYW5hcnkiLCJjbGllbnRfdmVyc2lvbiI6IjEuMC4yMSIsIm9zX3ZlcnNpb24iOiIxMC4wLjE5MDQyIiwib3NfYXJjaCI6Ing2NCIsInN5c3RlbV9sb2NhbGUiOiJlbi1VUyIsImNsaWVudF9idWlsZF9udW1iZXIiOjc1NjU3LCJjbGllbnRfZXZlbnRfc291cmNlIjpudWxsfQ==',
   xContextProperties: 'eyJsb2NhdGlvbiI6Ikludml0ZSBCdXR0b24gRW1iZWQiLCJsb2NhdGlvbl9ndWlsZF9pZCI6bnVsbCwibG9jYXRpb25fY2hhbm5lbF9pZCI6IjgwNzgyMTYwMDM1ODc5MzI1NiIsImxvY2F0aW9uX2NoYW5uZWxfdHlwZSI6MSwibG9jYXRpb25fbWVzc2FnZV9pZCI6IjgwNzgyMTY5ODUwMTA1MDQzOSJ9',
   webhookCantReach: 'Unable to reach your webhook.',
   invalidConfig: 'Invalid settings object, if you require more help visit https://github.com/slow/nitro-sniper#default-configuration for an example. Exiting process...',
   noMode: 'No sniper mode provided in settings, exiting process...',
   invalidMode: 'Invalid mode provided, visit https://github.com/slow/nitro-sniper#default-configuration for a list of modes. Exiting process...',
   noMain: 'No main account token provided, exiting process...',
   noAlts: 'No alt tokens were provided, exiting process...',
   invalidTokens: 'All tokens provided were invalid, exiting process...',
   userAgent: 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) discord/1.0.21 Chrome/83.0.4103.122 Electron/9.3.5 Safari/537.36',
   ready: (accounts, servers) => `Logged in with ${accounts} ${accounts > 1 ? 'accounts' : 'account'} and ${servers} servers.`,
   unknownResponse: (code, location, author, end, body) => `Unknown Response | ${body} | Code: ${chalk.bold(code)} | ${location} | ${author} | ${end}`,
   bothModeNoAlts: 'No alts were able to log in, trying to use only main account.',
   bothModeNoMain: "Couldn't log into main account meaning sniping is not possible. Exiting process...",
   noPaymentMethod: 'Main token does not have a billing source, some codes will not be sniped.',
   invalidWebhookType: 'Invalid Webhook type.',
   joinedServer: (invite, server, location, author, account, time) => `${account} joined ${server} after ${time} | Invite: ${invite} | ${location} | ${author}`,
   giveawayWon: (prize, hoster, guild, channel) => `Won giveaway of ${chalk.bold(prize)} from ${hoster} in (${guild} > #${channel}).`,
   failedGiveawayReact: (prize, hoster, guild, channel, timeTook) => `Failed to react to giveaway of ${chalk.bold(prize)} from ${hoster.tag} in (${guild} > #${channel}) after ${timeTook}.`,
   reactedGiveaway: (prize, hoster, guild, channel, timeTook) => `Reacted to giveaway of ${chalk.bold(prize)} from ${hoster} in (${guild} > #${channel}) after ${timeTook}.`,
   dmHosterSuccess: (hoster, prize, server, channel, timeTook) => `DMed ${hoster} after winning ${chalk.bold(prize)} in (${server} > #${channel}) after ${timeTook}.`,
   dmHosterFail: (hoster, prize, server, channel, timeTook) => `Couldn't DM ${hoster} after winning ${chalk.bold(prize)} in (${server} > #${channel}) after ${timeTook}.`,
   paymentMethodFail: (body) => `Unable to get billing source. ${body}`,
   invalidTokenOnSnipe: (code, from, author, end) => `Invalid Token | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   phinError: (err, code, from, author, end) => `${err} | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   unknownCode: (code, from, author, end) => `Invalid Code | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   alreadyRedeemedCode: (code, from, author, end) => `Already Redeemed | Code: ${chalk.bold(code)} | ${from} | ${author} | ${end}`,
   snipedCode: (code, type, from, author, end) => `Success | Code: ${chalk.bold(code)} | ${type} | ${from} | ${author} | ${end}`,
   duplicateFound: (code, location, author) => `Avoiding Duplicate | Code: ${chalk.bold(code)} | ${location} | ${author}`,
   webhookTypes: ['codeInvalid', 'codeAlreadyRedeemed', 'codeSuccess', 'giveawayEntered', 'giveawayWin', 'inviteJoin'],
   fields: {
      codeFail: (time, code, location) => [
         { key: 'Time Taken', value: time },
         { key: 'Code', value: code },
         { key: 'Location', value: location },
      ],
      codeSuccess: (time, type, code, location) => [
         { key: 'Time Taken', value: time, },
         { key: 'Type', value: type },
         { key: 'Code', value: code },
         { key: 'Location', value: location },
      ],
      giveawayEntered: (server, channel, timeTook, prize) => [
         { key: 'Location', value: `${server} > #${channel}` },
         { key: 'Prize', value: prize },
         { key: 'Time Waited', value: timeTook }
      ],
      giveawayWin: (server, channel, prize) => [
         { key: 'Location', value: `${server} > #${channel}` },
         { key: 'Prize', value: prize },
      ],
      inviteJoin: (location, invite, server, timeTook) => [
         { key: 'Location', value: location },
         { key: 'Server Joined', value: server },
         { key: 'Invite Used', value: invite },
         { key: 'Time Waited', value: timeTook }
      ]
   },
   titles: {
      codeInvalid: 'Invalid Code',
      codeAlreadyRedeemed: 'Already Redeemed Code',
      codeSuccess: 'Nitro Sniped',
      giveawayEntered: 'Giveaway Entered',
      giveawayWin: 'Giveaway Won',
      inviteJoin: 'Joined Server'
   },
   colors: {
      error: '#FF0000',
      success: '#41FC9F',
      warn: '#FFF000'
   },
   paymentSourceURL: 'https://discord.com/api/v6/users/@me/billing/payment-sources',
   redeemCodeURL: (code) => `https://discord.com/api/v6/entitlements/gift-codes/${code}/redeem`,
   inviteFetchURL: (invite) => `https://discord.com/api/v6/invites/${encodeURIComponent(invite)}?with_counts=true`,
   joinURL: (invite) => `https://discord.com/api/v6/invites/${invite}`
};